transformer_engine/common/permutation/permutation.cu

```diff
@@ -0,0 +1,296 @@
+  T *frag_load_store_ptr = reinterpret_cast<T *>(&frag_load_store);
+
+  static constexpr int kElementsPerAccess = 16 / sizeof(T);
+
+  const T *source_row_ptr = input_bwd + source_token * num_cols;
+  for (int i = tid * kElementsPerAccess; i < num_cols; i += blockDim.x * kElementsPerAccess) {
+    TCompute frag_src[kElementsPerAccess];
+
+    frag_load_store = __ldlu(reinterpret_cast<const float4 *>(source_row_ptr + i));
+
+    for (int e = 0; e < kElementsPerAccess; e++) frag_src[e] = TCompute(frag_load_store_ptr[e]);
+
+    int index = source_token;
+
+    for (int k = 0; k < topKTile; k++) {
+      if (k == num_topK) break;
+
+      int dest_row = row_id_map[index];
+      index += num_rows;
+
+      if (dest_row != -1) {
+        if (hasProb) {
+          for (int e = 0; e < kElementsPerAccess; e++)
+            frag_load_store_ptr[e] = T(frag_src[e] * s_prob[k]);
+        } else {
+          for (int e = 0; e < kElementsPerAccess; e++) frag_load_store_ptr[e] = T(frag_src[e]);
+        }
+
+        T *dest_row_ptr = act_grad + dest_row * num_cols;
+        *(float4 *)(dest_row_ptr + i) = frag_load_store;
+
+        if (hasProb) {
+          const T *input_fwd_ptr = input_fwd + dest_row * num_cols;
+
+          frag_load_store = __ldlu(reinterpret_cast<const float4 *>(input_fwd_ptr + i));
+
+          TCompute frag_input_fwd[kElementsPerAccess];
+          for (int e = 0; e < kElementsPerAccess; e++)
+            frag_input_fwd[e] = TCompute(frag_load_store_ptr[e]);
+
+          for (int e = 0; e < kElementsPerAccess; e++) {
+            accum[k] += float(frag_src[e] * frag_input_fwd[e]);
+          }
+        }
+      }
+    }
+  }
+
+  if (hasProb) {
+    for (int k = 0; k < topKTile; k++) {
+      if (k == num_topK) break;
+
+      for (int mask = 16; mask > 0; mask /= 2) {
+        accum[k] = accum[k] + __shfl_xor_sync(0xffffffff, accum[k], mask, 32);
+      }
+    }
+
+    if (tid == 0) {
+      for (int k = 0; k < topKTile; k++) {
+        if (k == num_topK) break;
+        prob_grad[source_token * num_topK + k] = accum[k];
+      }
+    }
+  }
+}
+
+template <typename T, bool FWD>
+void moe_permutation_launcher(const void *input_, void *output_, const int *sorted_row_id,
+                              int *row_id_map, const float *prob, const int num_rows,
+                              const int num_topK, const int num_cols, const int num_out_tokens,
+                              cudaStream_t stream, float *prob_grad, const void *input_fwd_) {
+  using TCompute = typename std::conditional<(std::is_same<T, __nv_fp8_e5m2>::value ||
+                                              std::is_same<T, __nv_fp8_e4m3>::value),
+                                             half, T>::type;
+
+  static constexpr int kElementsPerAccess = 16 / sizeof(T);
+
+  const T *input = reinterpret_cast<const T *>(input_);
+  T *output = reinterpret_cast<T *>(output_);
+  const T *input_fwd = reinterpret_cast<const T *>(input_fwd_);
+
+  if (FWD) {
```

Comment:
```suggestion
  if (IS_PERMUTE) {
```

Reply:
`FWD` was removed by split the original function into two functions for permute and unpermute.